contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,897 @@
+# TypeScript Coding Standards
+
+For our Javascript code style rules, refer to the [JavaScript Style Guide](STYLE.md). 
+
+You can also refer to [the TypeScript Handbook](https://www.typescriptlang.org/docs/handbook/intro.html) for more information on how to use TypeScript.
+
+## Organizing Type Definitions
+
+### File structure
+
+- Reusable type definitions, such as models (e.g. Report), must have their own file and be placed in a shared directory, e.g. `src/types/`.
+- Types specific to a single file should remain inside that file to keep their scope limited and their usage clear.
+- Types specific to a component/file with platform-specific variants should have their own file and be placed in the same folder.
+
+### File Extensions
+
+For TypeScript files containing JSX code, use the `.tsx` file extension. For other TypeScript code use the `.ts` file extension.
+
+Files containing only type declarations should **not** use the `.d.ts` extension. The only exception is when defining a common interface for platform-specific implementations (see below).
+
+### Platform-Specific Variants
+
+In most cases, the code written for this repo should be platform-independent. In such cases, each module should have a single file, `index.ts`, which defines the module's exports. There are, however, some cases in which a feature is intrinsically tied to the underlying platform. In such cases, the following file extensions can be used to export platform-specific code from a module:
+
+- `index.native.ts` - Native (iOS and Android)
+- `index.android.ts` - Android
+- `index.ios.ts` - iOS
+- `index.website.ts` - Web
+- `index.desktop.ts` - Desktop
+
+Note that `index.ts` should contain the default implementation, and only platform-specific implementations should be done in their respective files. i.e: If you have mobile-specific implementation in `index.native.ts`, then the desktop/web implementation can be contained in a shared `index.ts`.
+
+For each platform-specific module, create shared type definitions in a separate `types.ts` file and place it in the same folder. `types.ts` has to export shared types which are **compatible with all platform-specific implementations**. Declare component props, return types, and other common types in this file.
+
+```ts
+type ComponentProps = {
+    foo: string;
+    bar: number;
+};
+
+type Helpers = {
+    doSomething: () => void;
+};
+
+export type { ComponentProps, Helpers };
+```
+
+In case there is no default implementation, you have to create a `index.d.ts` declaration file. When importing the module from other files, TypeScript will automatically pick up the type definitions from `index.d.ts`. Be careful when defining `index.d.ts` as declaration files aren't checked by the TypeScript compiler (with `skipLibCheck: true` - [source](https://www.typescriptlang.org/tsconfig#skipLibCheck)).
```

Comment:
Apologies if this has been asked-and-answered, but I wonder if, instead of creating two patterns where one of them causes us to lose type-safety, we should instead enforce that every module has a default implementation (i.e: if you have `index.desktop.js`, `index.android.js`, and `index.ios.js`, then we would have `index.js` and know that it must be for web)

Reply:
Here is a long [discussion](https://docs.google.com/document/d/1YczjOMTEOdbWhsn0nEC2MzA31ME7OsNw-DYhIEJ4Hk8/edit?disco=AAAAzBdnj5M) in TS config doc regarding this topic.

TL;DR: 
We extracted two types of platform specific variants:
- With default implementation, only when it makes sense to have it (usually applies to more than one platform)
- No default implementation, each platform is different (like in `getPlatform` utility - [example](https://github.com/blazejkustra/App/commit/b936507771d97d9c71bd640d123fd7d0477b8742)). 

@fabioh8010 comment on platform specific variants without default implementation 👇 

> I think we don't have better options than using `.d.ts` files for this kind of situation. The `getPlatform` is a perfect example why I need this type of file because we don't have `index.ts` for it. An possible alternative would be to choose one of the platform-specific files and rename it to `index.ts`, but this is terrible in my opinion. Using `.d.ts` will be very beneficial for us.

Reply:
> instead of creating two patterns where one of them causes us to lose type-safety, we should instead enforce that every module has a default implementation

We don't really lose type-safety with declaration files as long as the name of the function is the same and types are shared from `types.ts` file and used in all implementations.

> if you have index.desktop.js, index.android.js, and index.ios.js, then we would have index.js and know that it must be for web

I have two problems with this approach:
1. It doesn't really eliminate type-safety problem that you mentioned with declaration files. What if each platform has a return type of literal strings: 'desktop' for desktop, 'ios' for iOS, 'android' for Android and 'web' for web. Rename `index.web.ts` to `index.ts`, now the return type is 'web'. When importing, type 'web' is inferred, but it should be either `string` or `'ios' | 'android' | 'desktop' | 'web'` in reality. That's why `types.ts` is needed and it resolves all these issues.
2. It is less readable what is the purpose of default implementation, I would have to check all remaining files to know that

Reply:
If I were to standardize one approach, I would opt for always requiring implementations for each of the four platforms and including an `index.d.ts` file with a common interface + `types.ts` with shared types.

The approach described in the doc is also fine with me 👍

Reply:
> But if we really want to standardize it, I'll opt for the approach @gedu mentioned: always use index.ts for only export purposes.

What if there is no default implementation? Do we need to implement `Button.d.ts` file then? I think files structure is less readable with this approach.

> This approach is what I see in component libraries and is what I believe is the standard.

Could you provide an example? @hayata-suenaga

Reply:
`index.ts` is usually used to export stuff in component libraries

- [`mui` Button component](https://github.com/mui/material-ui/blob/master/packages/mui-base/src/Button/index.ts)
- [`fluent-react-native` Button component (Microsoft)](https://github.com/microsoft/fluentui-react-native/blob/main/packages/components/Button/src/index.ts)

Those libraries are natively typed and also have `Button.types.ts` for `Button` component, for example.

However...

> What if there is no default implementation? Do we need to implement Button.d.ts file then?

I tested this, and no, it didn't work without default implementation (without `Button.ts`) 😢 and creating `Buttno.d.ts` and `index.ts` seems too much.

So let's go with our original plan 👍